File: AI_models/views/Mock_interview.py
from django.shortcuts import render
from django.shortcuts import render
from django.http import JsonResponse
import google.generativeai as genai
import json
from dotenv import load_dotenv
import os
from django.views.decorators.csrf import csrf_exempt


def mock_interview(request):
    return render(request,"mock_interview.html")

load_dotenv()

# Configure the API key
genai.configure(api_key=os.getenv("api_key"))
@csrf_exempt
def generate_questions(job_role, company_name):
    prompt = f"Generate 10-15 interview questions for a {job_role} role at {company_name}. Include a mix of technical, soft skills, and aptitude questions. Return only the questions in a numbered list format."
    model = genai.GenerativeModel("gemini-1.5-pro")
    response = model.generate_content(prompt)
    questions = [q.strip() for q in response.text.split("\n") if q.strip()]
    return questions if questions else ["No questions generated. Please try again."]

@csrf_exempt
def evaluate_answers(responses):
    prompt = "Evaluate the following interview responses and provide a performance summary with strengths and improvement suggestions:\n"
    for r in responses:
        prompt += f"Q: {r['question']}\nA: {r['answer']}\n"
    
    model = genai.GenerativeModel("gemini-1.5-pro")
    response = model.generate_content(prompt)

    evaluation_text = response.candidates[0].content.parts[0].text if response.candidates else "No evaluation generated."

    return evaluation_text



# API to generate questions
@csrf_exempt
def get_questions(request):
    if request.method == "POST":
        try:
            data = json.loads(request.body)
            job_role = data.get("job_role")
            company_name = data.get("company_name")

            if not job_role or not company_name:
                return JsonResponse({"error": "Missing job role or company name"}, status=400)

            # 🔹 Correcting function call
            questions = generate_questions(job_role, company_name)  

            return JsonResponse({"questions": questions})

        except json.JSONDecodeError:
            return JsonResponse({"error": "Invalid JSON format"}, status=400)
        except Exception as e:
            return JsonResponse({"error": f"Internal server error: {str(e)}"}, status=500)

    return JsonResponse({"error": "Invalid request method"}, status=400)

# API to evaluate responses
@csrf_exempt
def evaluate(request):
    if request.method == "POST":
        try:
            data = json.loads(request.body)
            responses = data.get("responses")

            if not responses:
                return JsonResponse({"error": "No responses provided"}, status=400)

            evaluation = evaluate_answers(responses)
            return JsonResponse({"evaluation": evaluation})

        except json.JSONDecodeError:
            return JsonResponse({"error": "Invalid JSON format"}, status=400)
        except Exception as e:
            return JsonResponse({"error": f"Internal server error: {str(e)}"}, status=500)

    return JsonResponse({"error": "Invalid request method"}, status=400)

File: static/js/mock_interview.js
let questions = [];
let currentQuestionIndex = 0;
let responses = [];

async function startInterview() {
    const jobRole = document.getElementById("job_role").value;
    const companyName = document.getElementById("company_name").value;

    if (!jobRole || !companyName) {
        alert("Please enter both job role and company name.");
        return;
    }

    const response = await fetch("generate_questions/", {
        // const response = await fetch("{% url 'get_questions'%}", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ job_role: jobRole, company_name: companyName })
    });

    const data = await response.json();
    questions = data.questions;
    console.log
    if (questions.length === 0 || questions[0] === "No questions generated. Please try again.") {
        alert("No questions generated. Please try again.");
        return;
    }

    document.getElementById("input-section").style.display = "none";
    document.getElementById("interview-section").style.display = "block";
    showNextQuestion();
}

function showNextQuestion() {
    if (currentQuestionIndex < questions.length) {
        document.getElementById("question").innerText = questions[currentQuestionIndex];
        document.getElementById("answer").value = "";
    } else {
        finishInterview();
    }
}

function submitAnswer() {
    const answer = document.getElementById("answer").value;
    responses.push({ question: questions[currentQuestionIndex], answer });

    currentQuestionIndex++;
    showNextQuestion();
}

function skipQuestion() {
    responses.push({ question: questions[currentQuestionIndex], answer: "Skipped" });

    currentQuestionIndex++;
    showNextQuestion();
}

async function finishInterview() {
    document.getElementById("interview-section").style.display = "none";

    const response = await fetch("evaluate/", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ responses })
    });

    const data = await response.json();

    // Directly setting the evaluation text (since it's a string, not JSON array)
    document.getElementById("evaluation-result").innerHTML = `<p>${data.evaluation.replace(/\n/g, "<br>")}</p>`;
    document.getElementById("evaluation-section").style.display = "block";
}